### DISCUSSION TITLE
Map url path to page file path at runtime in a middleware?

### DISCUSSION THREAD

**Discussion by fabb** [vercel/next.js#64118] [Help]
### Summary

Is there some functionality I can use to convert an incoming url path (e.g. `/this/is/some/request/path`) to a page file path (`/this/[...slug]`) at runtime in a middleware?

### Additional information

I need to do this in a middleware, not a react component.

**Comment by MarouaneBenbetka** [comment]
in nextjs you can do :
  const router = useRouter();
  const { slug } = router.query;

**Reply by fabb** [reply]
Ah, I forgot to mention that I need to do this in a middleware, not in a react component.

**Reply by MarouaneBenbetka** [reply]
You mean something like this :
```
import { NextResponse } from 'next/server';

export function middleware(request) {
  const url = request.nextUrl.clone();
  if (url.pathname.startsWith('/this/is')) {
    url.pathname = `/this/${url.pathname.split('/').slice(2).join('/')}`;
    return NextResponse.rewrite(url);
  }
  return NextResponse.next();
}
```
if this answer your question mark it as answered

**Reply by fabb** [reply]
Hm no not really. I need the exact file path of my dynamic route page file. E.g. `/this/[...slug]` when the path on disk is `/app/this/[...slug].tsx`.

**Reply by MarouaneBenbetka** [reply]
check this you can modify it according to you logic

```
import { NextResponse } from 'next/server';

export function middleware(request) {
  const url = request.nextUrl;
  
  // Extract the path from the URL
  const path = url.pathname;

  // Assuming all dynamic routes follow a specific pattern, 
  // you can transform the URL path into a file path.
  // This example directly maps the path to a file structure 
  // by adding the base directory and file extension.
  let filePath = `/app${path}.tsx`;
  
  // For dynamic routes, like /this/[...slug], you need to manually handle the mapping
  // If your URL structure is complex, this might involve regex 
  if (path.includes('/this/')) {
    filePath = `/app/this/[...slug].tsx`;
  }
  console.log(filePath);

  return NextResponse.next();
}
```

**Reply by fabb** [reply]
> you need to manually handle the mapping

This is exactly what I want to avoid. Next.js knows which files and dynamic routes exist, and can map incoming urls to them at runtime. I‘d line to avoid rebuilding that transformation and rather utilize the same transformation next.js does internally because it should work without updates when more pages are added later to the application.
